core/vm/instructions.go

```diff
@@ -991,9 +991,10 @@ func opSelfdestruct6780(pc *uint64, interpreter *EVMInterpreter, scope *ScopeCon
 		if !interpreter.evm.Accesses.TouchBasicData(contractAddr[:], false, scope.Contract.UseGas, false) {
 			return nil, ErrExecutionReverted
 		}
-		balanceIsZero := interpreter.evm.StateDB.GetBalance(contractAddr).Sign() == 0
 
-		if _, isPrecompile := interpreter.evm.precompile(beneficiaryAddr); !(isPrecompile && balanceIsZero) {
+		balanceIsZero := interpreter.evm.StateDB.GetBalance(contractAddr).Sign() == 0
+		_, isPrecompile := interpreter.evm.precompile(beneficiaryAddr)
+		if (!isPrecompile && !interpreter.evm.isSystemContract(beneficiaryAddr)) || !balanceIsZero {
```

Comment:
![image](https://github.com/user-attachments/assets/1a699293-1fe7-4661-86d6-30a0e6f92e57)

Reply:
nit: also define `isSystemContract` so that the condition is easier to read